### PR TITLE
Maintenance page

### PR DESCRIPTION
**Pull request by KristampsWong**
Since #27 PR, I marge to this branch. The Manage Cars, Add Cars, Edit Cars, Car Information are connected with Backend now.

### CODE REVIEW

File: code/src/app/api/carInfo/route.ts [hunk 1]
```diff
@@ -217,144 +166,105 @@ export async function GET(req: Request) {
...
-    // Check existing license being registered with another car
-    const existingLicense = await car.findOne({
-      // eslint-disable-next-line no-underscore-dangle
-      $and: [{ license }, { _id: { $ne: carObject._id } }],
-    });
+    const entries = Object.entries(data);
+    const nonEmptyOrNull = entries.filter(
+      ([, val]) => val !== '' && val !== null,
+    );
+    const filteredData = Object.fromEntries(nonEmptyOrNull);
```
Comment (KristampsWong) [inline code/src/app/api/carInfo/route.ts:186]: In order to remove any properties from the (frontend) data  object that have values of an empty string or null.

File: code/src/app/api/carInfo/route.ts [hunk 2]
```diff
@@ -217,144 +166,105 @@ export async function GET(req: Request) {
...
+
     if (
-      isMaintenanceTypeRequired(
-        mileageLastTireChange,
-        rotationMileageTire,
-        mileage,
-      )
+      (filteredData.dateNextOilChange && !filteredData.mileageLastOilChange) ||
+      (filteredData.dateNextTireChange && !filteredData.mileageLastTireChange)
     ) {
```
Comment (KristampsWong) [inline code/src/app/api/carInfo/route.ts:216]: 'If a new date for last maintenance change is provided, then a new mileage for the last maintenance change should also be provided.'
In order to make sure both mileage and date been provide.

File: code/src/app/api/carInfo/route.ts [hunk 3]
```diff
@@ -217,144 +166,105 @@ export async function GET(req: Request) {
...
+      return NextResponse.json(
+        {
+          message:
+            'If a new date for last maintenance change is provided, then a new mileage for the last maintenance change should also be provided.',
+        },
+        { status: 402 },
+      );
     }
+    const currentCarData = await car.findOne({ _id: filteredData.carId });
+    const updatedCarData = { ...currentCarData.toObject(), ...filteredData };
```
Comment (KristampsWong) [inline code/src/app/api/carInfo/route.ts:227]: Get previous mileage for DB. Becuase there is a situation, if user only changed mileage but not change maitenance info, the OilOverdue and TireOverDue would not update. So get previous mileage from db, and compare with current to check if overdue or not

File: code/src/app/api/carInfo/route.ts [hunk 4]
```diff
@@ -217,144 +166,105 @@ export async function GET(req: Request) {
...
 
-    // DB Update
-    await dbConnect();
+    const maintenanceStatus = determineMaintenanceStatus(
+      updatedCarData,
+      (filteredData as any).mileage,
+    );
+    Object.assign(filteredData, maintenanceStatus);
```
Comment (KristampsWong) [inline code/src/app/api/carInfo/route.ts:233]: Check if need maintenance

File: code/src/app/api/carInfo/route.ts [hunk 5]
```diff
@@ -217,144 +166,105 @@ export async function GET(req: Request) {
...
```
Comment (KristampsWong) [inline code/src/app/api/carInfo/route.ts:232]: Check if need maintenance

**Review by ncruz93 — CHANGES_REQUESTED** [review]
You changed the logic of the maintenance calculation I implemented completed. There should be a date of next scheduled change automatically calculated so that the calendar can display it. Overdue is only when the initial date calculated had passed and the maintenance change was not completed. I don't think you should have overwritten all of my work. Refactoring/Cleaning it up is fine, but this is completely different and it no longer works. How were you planning to use this to populate the calendar?

**Review by Alisa007 — APPROVED** [review]

**Pull request closed without merging**